xarray/core/dataset.py

```diff
@@ -7798,14 +7799,25 @@ def diff(
                     variables[name] = var.isel(slice_end) - var.isel(slice_start)
                 else:
                     variables[name] = var.isel(slice_new)
+                dims_found = True
             else:
                 variables[name] = var
 
         difference = self._replace_with_new_dims(variables, indexes=indexes)
 
         if n > 1:
-            return difference.diff(dim, n - 1)
+            return difference.diff(dim, n - 1, dims_found=dims_found)
         else:
+            if dims_found:
+                return difference
+            else:
+                from warnings import warn
```

Comment:
(these can be reduced to a single `if not dims_found`, and then it'll fall through to the `return difference` below)